File: src/main/java/me/centralhardware/znatoki/telegram/statistic/service/ReportService.java
package me.centralhardware.znatoki.telegram.statistic.service;

import lombok.RequiredArgsConstructor;
import me.centralhardware.znatoki.telegram.statistic.entity.postgres.Service;
import me.centralhardware.znatoki.telegram.statistic.mapper.postgres.*;
import me.centralhardware.znatoki.telegram.statistic.report.MonthReport;
import org.apache.commons.collections4.CollectionUtils;
import org.springframework.stereotype.Component;

import java.io.File;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;

@Component
@RequiredArgsConstructor
public class ReportService {

    private final ServiceMapper serviceMapper;
    private final UserMapper userMapper;
    private final ClientService clientService;
    private final ServicesMapper servicesMapper;
    private final OrganizationMapper organizationMapper;

    public List<File> getReportsCurrent(Long id){
        return getReport(serviceMapper::getCuurentMontTimes, id);
    }

    public List<File> getReportPrevious(Long id){
        return getReport(serviceMapper::getPrevMonthTimes, id);
    }

    private List<File> getReport(Function<Long,List<Service>> getTime, Long id){
        var times = getTime.apply(id);
        if (CollectionUtils.isEmpty(times)) return Collections.emptyList();

        var user =  Optional.ofNullable(userMapper.getById(times.get(0).getChatId()));
        return user.map(telegramUser -> telegramUser.getServices()
                .stream()
                .map(it -> {
                    var date = times.stream()
                            .filter(time -> time.getServiceId().equals(it))
                            .findFirst()
                            .map(Service::getDateTime)
                            .orElse(null);
                    if (date == null) return null;

                    return new MonthReport(userMapper.getById(id).getName(),
                            it, servicesMapper.getKeyById(it),
                            date,
                            organizationMapper.getReportFields(telegramUser.getOrganizationId()),
                            organizationMapper.getById(telegramUser.getOrganizationId()).getClientName()).generate(times);
                })
                .filter(Objects::nonNull)
                .toList()).orElse(Collections.emptyList());
    }

}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/service/ReportService.kt
package me.centralhardware.znatoki.telegram.statistic.service

import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import me.centralhardware.znatoki.telegram.statistic.entity.Payment
import me.centralhardware.znatoki.telegram.statistic.entity.Lesson
import me.centralhardware.znatoki.telegram.statistic.entity.SubjectId
import me.centralhardware.znatoki.telegram.statistic.mapper.PaymentMapper
import me.centralhardware.znatoki.telegram.statistic.mapper.LessonMapper
import me.centralhardware.znatoki.telegram.statistic.mapper.TutorMapper
import me.centralhardware.znatoki.telegram.statistic.report.MonthReport
import java.io.File

object ReportService {

    fun getReportsCurrent(id: Long): List<File> {
        return getReport(LessonMapper::getCurrentMontTimes, PaymentMapper::getCurrentMonthPayments, id)
    }

    fun getReportPrevious(id: Long): List<File> {
        return getReport(LessonMapper::getPrevMonthTimes, PaymentMapper::getPrevMonthPayments, id)
    }

    private fun getReport(getTime: (Long, SubjectId) -> List<Lesson>, getPayments: (Long, SubjectId) -> List<Payment>, id: Long): List<File> {
        val user =
            TutorMapper.findByIdOrNull(id)?.let { user ->
                user.subjects.mapNotNull { subjectId ->
                    runBlocking {
                        val times = async { getTime.invoke(id, subjectId) }
                        val payments = async { getPayments.invoke(id, subjectId) }

                        val service = times.await().firstOrNull { time -> time.subjectId == subjectId }
                        if (service == null) return@runBlocking null

                        MonthReport(user.name, subjectId, id).generate(times.await(), payments.await())
                    }
                }
            }

        return user ?: emptyList()
    }
}
